File: apps/website/src/lib/editor-ffi/required-surface-pages.test.ts
import { describe, expect, it } from 'vitest';
import { requiredSurfacePages } from './required-surface-pages';
import { resolveInstantRevealPreparationViewports, resolveNearestScrollTop, resolveTypewriterScrollTop } from './scroll';
import type { SurfacePageSpan, VerticalSpan } from './required-surface-pages';

const pages = (...boundaries: number[]): SurfacePageSpan[] =>
  boundaries.slice(0, -1).map((top, page) => ({ page, top, bottom: boundaries[page + 1] }));

const required = ({
  pages,
  currentViewport = null,
  activePages = new Set(),
  preparationViewports = [],
}: {
  pages: SurfacePageSpan[];
  currentViewport?: VerticalSpan | null;
  activePages?: Set<number>;
  preparationViewports?: VerticalSpan[];
}): number[] => [...requiredSurfacePages({ pages, currentViewport, activePages, preparationViewports })].toSorted((a, b) => a - b);

describe('requiredSurfacePages', () => {
  it('empty-document', () => {
    expect(
      required({
        pages: [],
        currentViewport: { top: 0, bottom: 100 },
        activePages: new Set([0]),
        preparationViewports: [{ top: 0, bottom: 100 }],
      }),
    ).toEqual([]);
  });

  it('ignores invalid geometry', () => {
    const pageSpans = pages(0, 100, 200);
    expect(
      required({
        pages: pageSpans,
        currentViewport: { top: NaN, bottom: 100 },
        activePages: new Set([1]),
        preparationViewports: [{ top: 300, bottom: 200 }],
      }),
    ).toEqual([]);
  });

  it('current-viewport-acquire-one-height', () => {
    expect(required({ pages: pages(0, 100, 200, 300, 400, 500), currentViewport: { top: 200, bottom: 300 } })).toEqual([1, 2, 3]);
  });

  it('active-target-release-one-and-a-half-heights', () => {
    const pageSpans = [
      { page: 0, top: 0, bottom: 50 },
      { page: 1, top: 50, bottom: 100 },
      { page: 2, top: 100, bottom: 200 },
      { page: 3, top: 200, bottom: 300 },
      { page: 4, top: 300, bottom: 400 },
      { page: 5, top: 400, bottom: 450 },
      { page: 6, top: 450, bottom: 500 },
    ];

    expect(
      required({
        pages: pageSpans,
        currentViewport: { top: 200, bottom: 300 },
        activePages: new Set([0, 1, 5, 6]),
      }),
    ).toEqual([1, 2, 3, 4, 5]);
  });

  it('preparation viewport uses the same acquire range as the destination viewport', () => {
    const pageSpans = [
      { page: 0, top: 0, bottom: 100 },
      { page: 1, top: 120, bottom: 220 },
      { page: 2, top: 240, bottom: 340 },
    ];

    expect(required({ pages: pageSpans, preparationViewports: [{ top: 100, bottom: 120 }] })).toEqual([0, 1]);
    expect(required({ pages: pageSpans, preparationViewports: [{ top: 99, bottom: 120 }] })).toEqual([0, 1]);
    expect(required({ pages: pageSpans, preparationViewports: [{ top: 100, bottom: 121 }] })).toEqual([0, 1]);
  });

  it('zoomed-origin-and-gaps', () => {
    const alreadyDerivedPageSpans = [
      { page: 0, top: 50, bottom: 250 },
      { page: 1, top: 290, bottom: 490 },
    ];

    expect(required({ pages: alreadyDerivedPageSpans, preparationViewports: [{ top: 245, bottom: 295 }] })).toEqual([0, 1]);
  });

  it('out-of-range-active-pages', () => {
    expect(
      required({
        pages: pages(0, 100, 200, 300),
        currentViewport: { top: 0, bottom: 100 },
        activePages: new Set([-1, 2, 3, 100]),
      }),
    ).toEqual([0, 1, 2]);
  });

  it('one-to-three-append', () => {
    const currentViewport = { top: 100, bottom: 200 };

    expect(required({ pages: pages(0, 100), currentViewport, activePages: new Set([0]) })).toEqual([0]);
    expect(required({ pages: pages(0, 100, 200, 300), currentViewport, activePages: new Set([0]) })).toEqual([0, 1, 2]);
  });

  it('four-to-one-shrink', () => {
    expect(
      required({
        pages: pages(0, 100),
        currentViewport: { top: 100, bottom: 200 },
        activePages: new Set([0, 1, 2, 3]),
      }),
    ).toEqual([0]);
  });

  it('long-document-bounded-count', () => {
    const pageSpans = Array.from({ length: 100 }, (_, page) => ({ page, top: page * 100, bottom: (page + 1) * 100 }));

    expect(required({ pages: pageSpans, currentViewport: { top: 5000, bottom: 5100 } })).toEqual([49, 50, 51]);
  });

  it('disjoint-preparation-viewports-do-not-fill-gaps', () => {
    const pageSpans = Array.from({ length: 10 }, (_, page) => ({ page, top: page * 100, bottom: (page + 1) * 100 }));

    expect(
      required({
        pages: pageSpans,
        preparationViewports: [
          { top: 210, bottom: 220 },
          { top: 810, bottom: 820 },
        ],
      }),
    ).toEqual([2, 8]);
  });

  it('instant preparation covers the first destination viewport requirements', () => {
    const pageSpans = pages(0, 100, 200, 300, 400, 500, 600);
    const destinationViewport = { top: 300, bottom: 400 };
    const prepared = required({
      pages: pageSpans,
      currentViewport: { top: 0, bottom: 100 },
      activePages: new Set([0, 1]),
      preparationViewports: [destinationViewport],
    });
    const firstDestinationRequirements = required({
      pages: pageSpans,
      currentViewport: destinationViewport,
      activePages: new Set(prepared),
    });

    expect(prepared).toEqual(expect.arrayContaining(firstDestinationRequirements));
  });

  it('instant preparation contains every destination supported by the production scroll resolvers', () => {
    const metrics = {
      clientHeight: 400,
      scrollHeight: 1400,
      targetTop: 500,
      targetBottom: 650,
      visibleArea: { topInset: 10, bottomInset: 20 },
    };
    const preparation = resolveInstantRevealPreparationViewports({
      mode: 'nearest',
      scrollTop: 350,
      ...metrics,
    });
    const exactDestinations = [0, 350, 800].map((scrollTop) => resolveNearestScrollTop({ scrollTop, ...metrics }) ?? scrollTop);

    for (const scrollTop of exactDestinations) {
      expect(preparation).toContainEqual({ top: scrollTop, bottom: scrollTop + metrics.clientHeight });
    }
  });

  it('instant preparation covers a target taller than the visible range', () => {
    const metrics = {
      clientHeight: 400,
      scrollHeight: 2000,
      targetTop: 600,
      targetBottom: 1100,
      visibleArea: { topInset: 20, bottomInset: 30 },
    };
    const preparation = resolveInstantRevealPreparationViewports({ mode: 'nearest', scrollTop: 0, ...metrics });

    for (const scrollTop of [0, 600, 1500]) {
      const destination = resolveNearestScrollTop({ scrollTop, ...metrics });
      expect(destination).not.toBeNull();
      expect(preparation).toContainEqual({ top: destination, bottom: Number(destination) + metrics.clientHeight });
    }
  });

  it('instant preparation uses production clamp centered fallback and typewriter alignment', () => {
    expect(
      resolveInstantRevealPreparationViewports({
        mode: 'nearest',
        scrollTop: 200,
        clientHeight: 400,
        scrollHeight: 650,
        targetTop: 500,
        targetBottom: 650,
        visibleArea: { topInset: 10, bottomInset: 20 },
      }),
    ).toEqual([{ top: 250, bottom: 650 }]);

    const centeredMetrics = {
      scrollTop: 0,
      clientHeight: 180,
      scrollHeight: 1180,
      targetTop: 500,
      targetBottom: 520,
      visibleArea: { topInset: 70, bottomInset: 70 },
    };
    const centeredDestination = resolveNearestScrollTop(centeredMetrics);
    expect(centeredDestination).not.toBeNull();
    expect(resolveInstantRevealPreparationViewports({ mode: 'nearest', ...centeredMetrics })).toEqual([
      { top: centeredDestination, bottom: Number(centeredDestination) + centeredMetrics.clientHeight },
    ]);

    const typewriterMetrics = {
      scrollTop: 0,
      clientHeight: 400,
      scrollHeight: 1900,
      targetTop: 800,
      targetBottom: 820,
      position: 0.5,
    };
    const typewriterDestination = resolveTypewriterScrollTop(typewriterMetrics);
    expect(resolveInstantRevealPreparationViewports({ mode: 'typewriter', ...typewriterMetrics })).toEqual([
      { top: typewriterDestination, bottom: Number(typewriterDestination) + typewriterMetrics.clientHeight },
    ]);
  });

  it('bounded instant preparation covers each destination cohort without filling disjoint gaps', () => {
    const pageSpans = Array.from({ length: 30 }, (_, page) => ({ page, top: page * 100, bottom: (page + 1) * 100 }));
    const preparationViewports = [
      { targetTop: 300, targetBottom: 320 },
      { targetTop: 2300, targetBottom: 2320 },
    ].flatMap((target) =>
      resolveInstantRevealPreparationViewports({
        mode: 'typewriter',
        scrollTop: 0,
        clientHeight: 100,
        scrollHeight: 3000,
        position: 0.5,
        ...target,
      }),
    );
    const prepared = required({
      pages: pageSpans,
      currentViewport: { top: 0, bottom: 100 },
      preparationViewports,
    });

    for (const destination of preparationViewports) {
      const destinationRequired = required({ pages: pageSpans, currentViewport: destination });
      expect(prepared).toEqual(expect.arrayContaining(destinationRequired));
    }
    expect(prepared).not.toContain(12);
  });

  it('offscreen-selection-is-not-implicit-demand', () => {
    const pageSpans = Array.from({ length: 10 }, (_, page) => ({ page, top: page * 100, bottom: (page + 1) * 100 }));

    // There is intentionally no selection input that could add the offscreen page.
    const result = required({ pages: pageSpans, currentViewport: { top: 0, bottom: 100 } });
    expect(result).toEqual([0, 1]);
    expect(result).not.toContain(8);
  });
});

File: apps/mobile/compose/src/commonTest/kotlin/co/typie/editor/RequiredSurfacePagesTest.kt
package co.typie.editor

import androidx.compose.ui.geometry.Size
import co.typie.editor.scroll.EditorBringIntoViewPolicy
import co.typie.editor.scroll.EditorVisibleArea
import co.typie.editor.scroll.resolveEditorAutoScrollPolicy
import co.typie.editor.scroll.resolveInstantRevealPreparationViewports
import co.typie.editor.scroll.resolveKeepVisibleScrollOffset
import co.typie.editor.scroll.resolveTypewriterScrollOffset
import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertFalse
import kotlin.test.assertTrue

class RequiredSurfacePagesTest {
  @Test
  fun `empty-document`() {
    assertEquals(
      emptySet(),
      requiredSurfacePages(
        pages = emptyList(),
        currentViewport = VerticalSpan(0f, 100f),
        activePages = setOf(0),
        preparationViewports = listOf(VerticalSpan(0f, 100f)),
      ),
    )
  }

  @Test
  fun `ignores invalid geometry`() {
    val pages = pages(0f, 100f, 200f)

    assertEquals(
      emptySet(),
      requiredSurfacePages(
        pages = pages,
        currentViewport = VerticalSpan(Float.NaN, 100f),
        activePages = setOf(1),
        preparationViewports = listOf(VerticalSpan(300f, 200f)),
      ),
    )
  }

  @Test
  fun `current-viewport-acquire-one-height`() {
    assertEquals(
      setOf(1, 2, 3),
      requiredSurfacePages(
        pages = pages(0f, 100f, 200f, 300f, 400f, 500f),
        currentViewport = VerticalSpan(200f, 300f),
      ),
    )
  }

  @Test
  fun `active-target-release-one-and-a-half-heights`() {
    val pages =
      listOf(
        SurfacePageSpan(page = 0, top = 0f, bottom = 50f),
        SurfacePageSpan(page = 1, top = 50f, bottom = 100f),
        SurfacePageSpan(page = 2, top = 100f, bottom = 200f),
        SurfacePageSpan(page = 3, top = 200f, bottom = 300f),
        SurfacePageSpan(page = 4, top = 300f, bottom = 400f),
        SurfacePageSpan(page = 5, top = 400f, bottom = 450f),
        SurfacePageSpan(page = 6, top = 450f, bottom = 500f),
      )

    assertEquals(
      setOf(1, 2, 3, 4, 5),
      requiredSurfacePages(
        pages = pages,
        currentViewport = VerticalSpan(200f, 300f),
        activePages = setOf(0, 1, 5, 6),
      ),
    )
  }

  @Test
  fun `preparation viewport uses the same acquire range as the destination viewport`() {
    val pages =
      listOf(
        SurfacePageSpan(page = 0, top = 0f, bottom = 100f),
        SurfacePageSpan(page = 1, top = 120f, bottom = 220f),
        SurfacePageSpan(page = 2, top = 240f, bottom = 340f),
      )

    assertEquals(
      setOf(0, 1),
      requiredSurfacePages(pages = pages, preparationViewports = listOf(VerticalSpan(100f, 120f))),
    )
    assertEquals(
      setOf(0, 1),
      requiredSurfacePages(pages = pages, preparationViewports = listOf(VerticalSpan(99f, 120f))),
    )
    assertEquals(
      setOf(0, 1),
      requiredSurfacePages(pages = pages, preparationViewports = listOf(VerticalSpan(100f, 121f))),
    )
  }

  @Test
  fun `zoomed-origin-and-gaps`() {
    val alreadyDerivedPageSpans =
      listOf(
        SurfacePageSpan(page = 0, top = 50f, bottom = 250f),
        SurfacePageSpan(page = 1, top = 290f, bottom = 490f),
      )

    assertEquals(
      setOf(0, 1),
      requiredSurfacePages(
        pages = alreadyDerivedPageSpans,
        preparationViewports = listOf(VerticalSpan(245f, 295f)),
      ),
    )
  }

  @Test
  fun `out-of-range-active-pages`() {
    assertEquals(
      setOf(0, 1, 2),
      requiredSurfacePages(
        pages = pages(0f, 100f, 200f, 300f),
        currentViewport = VerticalSpan(0f, 100f),
        activePages = setOf(-1, 2, 3, 100),
      ),
    )
  }

  @Test
  fun `one-to-three-append`() {
    val viewport = VerticalSpan(100f, 200f)

    assertEquals(
      setOf(0),
      requiredSurfacePages(
        pages = pages(0f, 100f),
        currentViewport = viewport,
        activePages = setOf(0),
      ),
    )
    assertEquals(
      setOf(0, 1, 2),
      requiredSurfacePages(
        pages = pages(0f, 100f, 200f, 300f),
        currentViewport = viewport,
        activePages = setOf(0),
      ),
    )
  }

  @Test
  fun `four-to-one-shrink`() {
    assertEquals(
      setOf(0),
      requiredSurfacePages(
        pages = pages(0f, 100f),
        currentViewport = VerticalSpan(100f, 200f),
        activePages = setOf(0, 1, 2, 3),
      ),
    )
  }

  @Test
  fun `long-document-bounded-count`() {
    val pages = (0 until 100).map { page -> SurfacePageSpan(page, page * 100f, (page + 1) * 100f) }

    assertEquals(
      setOf(49, 50, 51),
      requiredSurfacePages(pages = pages, currentViewport = VerticalSpan(5_000f, 5_100f)),
    )
  }

  @Test
  fun `disjoint-preparation-viewports-do-not-fill-gaps`() {
    val pages = (0 until 10).map { page -> SurfacePageSpan(page, page * 100f, (page + 1) * 100f) }

    assertEquals(
      setOf(2, 8),
      requiredSurfacePages(
        pages = pages,
        preparationViewports = listOf(VerticalSpan(210f, 220f), VerticalSpan(810f, 820f)),
      ),
    )
  }

  @Test
  fun `instant preparation covers the first destination viewport requirements`() {
    val pageSpans = pages(0f, 100f, 200f, 300f, 400f, 500f, 600f)
    val destinationViewport = VerticalSpan(300f, 400f)
    val prepared =
      requiredSurfacePages(
        pages = pageSpans,
        currentViewport = VerticalSpan(0f, 100f),
        activePages = setOf(0, 1),
        preparationViewports = listOf(destinationViewport),
      )
    val firstDestinationRequirements =
      requiredSurfacePages(
        pages = pageSpans,
        currentViewport = destinationViewport,
        activePages = prepared,
      )

    assertTrue(
      prepared.containsAll(firstDestinationRequirements),
      "preparation=$prepared destination=$firstDestinationRequirements",
    )
  }

  @Test
  fun `instant preparation contains every destination supported by the production scroll resolvers`() {
    val visibleArea =
      EditorVisibleArea(
        viewport = Size(width = 800f, height = 400f),
        topInset = 10f,
        bottomOcclusionInset = 20f,
      )
    val target = VerticalSpan(top = 500f, bottom = 650f)
    val preparation =
      resolveInstantRevealPreparationViewports(
        currentScroll = 350f,
        viewportHeight = 400f,
        maximumScrollY = 1_000f,
        target = target,
        visibleArea = visibleArea,
        autoScrollPolicy = resolveEditorAutoScrollPolicy(visibleArea),
        policy = EditorBringIntoViewPolicy.CursorGuard,
      )

    val exactDestinations =
      listOf(0f, 350f, 800f).map { currentScroll ->
        resolveKeepVisibleScrollOffset(
          currentScroll = currentScroll,
          targetTopInContent = target.top,
          targetBottomInContent = target.bottom,
          visibleArea = visibleArea,
        ) ?: currentScroll
      }

    exactDestinations.forEach { scrollY ->
      assertTrue(
        VerticalSpan(scrollY, scrollY + visibleArea.viewport.height) in preparation,
        "destination=$scrollY preparation=$preparation",
      )
    }
  }

  @Test
  fun `instant preparation covers a target taller than the visible range`() {
    val visibleArea =
      EditorVisibleArea(
        viewport = Size(width = 800f, height = 400f),
        topInset = 20f,
        bottomOcclusionInset = 30f,
      )
    val target = VerticalSpan(top = 600f, bottom = 1_100f)
    for (currentScroll in listOf(0f, 600f, 1_500f)) {
      val preparation =
        resolveInstantRevealPreparationViewports(
          currentScroll = currentScroll,
          viewportHeight = visibleArea.viewport.height,
          maximumScrollY = 1_600f,
          target = target,
          visibleArea = visibleArea,
          autoScrollPolicy = resolveEditorAutoScrollPolicy(visibleArea),
          policy = EditorBringIntoViewPolicy.CursorGuard,
        )
      val destination =
        resolveKeepVisibleScrollOffset(
          currentScroll = currentScroll,
          targetTopInContent = target.top,
          targetBottomInContent = target.bottom,
          visibleArea = visibleArea,
        ) ?: currentScroll
      assertTrue(
        VerticalSpan(destination, destination + visibleArea.viewport.height) in preparation,
        "destination=$destination preparation=$preparation",
      )
    }
  }

  @Test
  fun `instant preparation uses production clamp centered fallback and typewriter alignment`() {
    val clamped =
      resolveInstantRevealPreparationViewports(
        currentScroll = 200f,
        viewportHeight = 400f,
        maximumScrollY = 250f,
        target = VerticalSpan(top = 500f, bottom = 650f),
        visibleArea =
          EditorVisibleArea(
            viewport = Size(width = 800f, height = 400f),
            topInset = 10f,
            bottomOcclusionInset = 20f,
          ),
        autoScrollPolicy =
          resolveEditorAutoScrollPolicy(
            EditorVisibleArea(
              viewport = Size(width = 800f, height = 400f),
              topInset = 10f,
              bottomOcclusionInset = 20f,
            )
          ),
        policy = EditorBringIntoViewPolicy.CursorGuard,
      )
    assertEquals(setOf(VerticalSpan(250f, 650f)), clamped.toSet())

    val narrowVisibleArea =
      EditorVisibleArea(
        viewport = Size(width = 800f, height = 180f),
        topInset = 70f,
        bottomOcclusionInset = 70f,
      )
    val centeredTarget = VerticalSpan(top = 500f, bottom = 520f)
    val centered =
      resolveInstantRevealPreparationViewports(
        currentScroll = 0f,
        viewportHeight = 180f,
        maximumScrollY = 1_000f,
        target = centeredTarget,
        visibleArea = narrowVisibleArea,
        autoScrollPolicy = resolveEditorAutoScrollPolicy(narrowVisibleArea),
        policy = EditorBringIntoViewPolicy.CursorGuard,
      )
    val centeredDestination =
      requireNotNull(
        resolveKeepVisibleScrollOffset(
          currentScroll = 0f,
          targetTopInContent = centeredTarget.top,
          targetBottomInContent = centeredTarget.bottom,
          visibleArea = narrowVisibleArea,
        )
      )
    assertEquals(listOf(VerticalSpan(centeredDestination, centeredDestination + 180f)), centered)

    val typewriterVisibleArea = EditorVisibleArea(viewport = Size(width = 800f, height = 400f))
    val typewriterTarget = VerticalSpan(top = 800f, bottom = 820f)
    val typewriter =
      resolveInstantRevealPreparationViewports(
        currentScroll = 0f,
        viewportHeight = 400f,
        maximumScrollY = 1_500f,
        target = typewriterTarget,
        visibleArea = typewriterVisibleArea,
        autoScrollPolicy =
          resolveEditorAutoScrollPolicy(
            visibleArea = typewriterVisibleArea,
            typewriterEnabled = true,
            typewriterPosition = 0.5f,
          ),
        policy = EditorBringIntoViewPolicy.Typewriter,
      )
    val typewriterDestination =
      requireNotNull(
        resolveTypewriterScrollOffset(
          currentScroll = 0f,
          targetTopInContent = typewriterTarget.top,
          targetBottomInContent = typewriterTarget.bottom,
          visibleArea = typewriterVisibleArea,
          position = 0.5f,
        )
      )
    assertEquals(
      listOf(VerticalSpan(typewriterDestination, typewriterDestination + 400f)),
      typewriter,
    )

    val oversizedTypewriterTarget = VerticalSpan(top = 600f, bottom = 1_100f)
    val oversizedTypewriter =
      resolveInstantRevealPreparationViewports(
        currentScroll = 0f,
        viewportHeight = 400f,
        maximumScrollY = 1_600f,
        target = oversizedTypewriterTarget,
        visibleArea =
          EditorVisibleArea(
            viewport = Size(width = 800f, height = 400f),
            topInset = 20f,
            bottomOcclusionInset = 30f,
          ),
        autoScrollPolicy =
          resolveEditorAutoScrollPolicy(
            visibleArea =
              EditorVisibleArea(
                viewport = Size(width = 800f, height = 400f),
                topInset = 20f,
                bottomOcclusionInset = 30f,
              ),
            typewriterEnabled = true,
            typewriterPosition = 0.5f,
          ),
        policy = EditorBringIntoViewPolicy.Typewriter,
      )
    assertEquals(
      listOf(VerticalSpan(0f, 400f), VerticalSpan(520f, 920f), VerticalSpan(790f, 1_190f)),
      oversizedTypewriter,
    )

    val documentEdgeTarget = VerticalSpan(top = 10f, bottom = 500f)
    val documentEdge =
      resolveInstantRevealPreparationViewports(
        currentScroll = 800f,
        viewportHeight = 400f,
        maximumScrollY = 1_600f,
        target = documentEdgeTarget,
        visibleArea =
          EditorVisibleArea(
            viewport = Size(width = 800f, height = 400f),
            topInset = 20f,
            bottomOcclusionInset = 30f,
          ),
        autoScrollPolicy =
          resolveEditorAutoScrollPolicy(
            visibleArea =
              EditorVisibleArea(
                viewport = Size(width = 800f, height = 400f),
                topInset = 20f,
                bottomOcclusionInset = 30f,
              ),
            typewriterEnabled = true,
            typewriterPosition = 0.5f,
          ),
        policy = EditorBringIntoViewPolicy.Typewriter,
      )
    assertEquals(
      listOf(VerticalSpan(800f, 1_200f), VerticalSpan(0f, 400f), VerticalSpan(190f, 590f)),
      documentEdge,
    )
  }

  @Test
  fun `bounded instant preparation covers each destination cohort without filling disjoint gaps`() {
    val pageSpans =
      (0 until 30).map { page -> SurfacePageSpan(page, page * 100f, (page + 1) * 100f) }
    val visibleArea = EditorVisibleArea(viewport = Size(width = 800f, height = 100f))
    val preparationViewports =
      listOf(VerticalSpan(300f, 320f), VerticalSpan(2_300f, 2_320f)).flatMap { target ->
        resolveInstantRevealPreparationViewports(
          currentScroll = 0f,
          viewportHeight = 100f,
          maximumScrollY = 2_900f,
          target = target,
          visibleArea = visibleArea,
          autoScrollPolicy =
            resolveEditorAutoScrollPolicy(
              visibleArea = visibleArea,
              typewriterEnabled = true,
              typewriterPosition = 0.5f,
            ),
          policy = EditorBringIntoViewPolicy.Typewriter,
        )
      }
    val prepared =
      requiredSurfacePages(
        pages = pageSpans,
        currentViewport = VerticalSpan(0f, 100f),
        preparationViewports = preparationViewports,
      )

    preparationViewports.forEach { destination ->
      val destinationRequired =
        requiredSurfacePages(pages = pageSpans, currentViewport = destination)
      assertTrue(
        prepared.containsAll(destinationRequired),
        "destination=$destination prepared=$prepared required=$destinationRequired",
      )
    }
    assertFalse(
      12 in prepared,
      "disjoint preparation must not acquire pages between destinations: $prepared",
    )
  }

  @Test
  fun `offscreen-selection-is-not-implicit-demand`() {
    val required =
      requiredSurfacePages(
        pages = (0 until 10).map { page -> SurfacePageSpan(page, page * 100f, (page + 1) * 100f) },
        currentViewport = VerticalSpan(0f, 100f),
      )

    // There is intentionally no selection input that could add the offscreen page.
    assertEquals(setOf(0, 1), required)
    assertFalse(8 in required)
  }
}

private fun pages(vararg boundaries: Float): List<SurfacePageSpan> =
  boundaries.asList().zipWithNext().mapIndexed { page, (top, bottom) ->
    SurfacePageSpan(page, top, bottom)
  }
